docs/design/proposals/native-binding/native-binding.md

```diff
@@ -0,0 +1,150 @@
+
+## Requirements
+
+1. Export any number of Spec and/or Status values to application consumable formats (ConfigMap and Secrets)
+2. Export to any namespace (for cluster-scoped controllers only)
+3. The exported value should stay in sync as the resource changes
+4. Resources should not need to be updated to add/remove bindings
+
+## Existing Solutions
+
+### Redhat Service Binding Operator
+
+>“[Redhat Service Binding Operator] enables developers to connect their application to backing services with a consistent and predictable experience“ - https://github.com/redhat-developer/service-binding-operator
+
+
+The [Redhat Service Binding Operator](https://github.com/redhat-developer/service-binding-operator) (herein called the SBO) uses annotations present on CRs and resources to import field values from other resources. The main purpose of SBO is to standardise the system for injecting secret information from database services into application deployment environment variables. The SBO can automatically detect and bind to services only from a limited number of [supported operators](https://github.com/redhat-developer/service-binding-operator#known-bindable-operators). 
+
+SBO is not suitable for this design as it can only support a limited number of fields based on pre-configured database bindings. That is, it can handle the case where an application developer wants to inject a PostgreSQL database connection string into a deployment, but little more than that. SBO does not support exporting to configmap or secrets directly.
+
+## Proposed Solution
+
+### ACK Binding CRD
+
+ACK uses a common CRD to support resource adoption, implemented through a separate (common) reconciler loop in each controller. Common CRDs allow customers to define controller configuration for tasks that do not fit within the spec of existing CRDs or that are too complex to fit into the annotations. 
+
+This solution would introduce a new CRD solely for the purpose of declaring field exports. The CRD spec would contain an array of structs, each containing:
+
+* The source path, containing:
+    * Resource API group, kind, namespace, name
+* Spec/Status field path
+* An output paths, containing:
+    * Type: ConfigMap or Secret
+    * Namespace and name
+    * (Optionally) Output structure
+
+The common field exporter reconciler running within the controller would have a shared informer for all custom resources of this type. The reconciler would read the spec for each binding resource, filter for any resources where the source path matches its respective resource API group, version and kind and then read and export the corresponding fields. 
+
+Having this configuration as a separate CRD allows application developers to specify their own configuration independent of the ops teams that manage the infrastructure. Cluster operators can lock down application teams’ RBAC permissions to only have access to read ACK resources, with an exception only to write this new binding CRD.
```

Comment:
> Couldn't the application developer circumvent the lack of `read` permissions on the CR by simply creating a bunch of `FieldExport` CRs that essentially exported all fields in the CR? :(

Yeah ... this did come up in the design review. We really didn't want to go down the route of reinventing an allow-/block-listing configuration because that would basically be reimplementing RBAC. Instead, we decided to limit this design so that you can only read fields from resources in the same namespace. We assume that if the developer has been given access to a namespace that they could be allowed to read anything within it. 

> I actually don't think it's going to be common for cluster operators to give write permission on the FieldExport CR when they don't give read permissions to the application developer role to begin with

I agree. My use case above addresses the case where the cluster ops team doesn't permit *write* access. In which case the application developer couldn't patch an annotation onto the resource to configure the exporting - hence requiring a separate CR. I am assuming for most of this that the app dev has read access.

Something that I haven't put in this doc that I just thought about is how to handle exporting fields that have secrets attached. In my mind we would not allow anyone to export those fields. Partly because it would require resolving the field and doing a lookup, which requires more RBAC permissions. Mostly though because if you need that secret in your application, you should already be able to just mount the secret as an environment variable directly, rather than exporting it and then mounting it.